Popularity/DCAFPilot/src/python/DCAF/ml/model.py

```diff
@@ -198,24 +207,84 @@ def model(train_file, newdata_file, idcol, tcol, learner, lparams=None,
-        dbses = [int(i) for i in list(tdf[dbs_h])]
-        if  scaler:
-            tdf = getattr(preprocessing, scaler)().fit_transform(tdf)
-        predictions = fit.predict(tdf)
-        data = {'dataset':datasets, dbs_h: dbses, 'prediction':predictions}
-        out = pd.DataFrame(data=data)
-        if  ofile:
-            out.to_csv(ofile, header=True, index=False)
+        nfiles = []
+        if  os.path.isfile(newdata_file):
+            nfiles = [newdata_file]
+        else:
+            def idcheck():
+                if  ofile.find('(id)') == -1:
+                    msg  = "ERROR: since file list for --newdata is provided, "
+                    msg += "--predict has to contain string '(id)', see model --help"
+                    print(msg)
+                    sys.exit(1)
+                else:
+                    return True
+            if newdata_file.find(',') != -1 and idcheck():
+                nfiles = newdata_file.split(',')
+            elif newdata_file.find('*') != -1 and idcheck():
+                nfiles = glob.glob(newdata_file)
+            elif os.path.isdir(newdata_file) and idcheck():
+                for ext in ['.csv.gz', '.csv', 'csv.bz2']:
+                    nfiles = [f for f in findfiles(fin, ext)]
+                    if  not len(nfiles):
+                        print("WARNING: no files to predict in %s" % newdata_file)
+                        return
+            else:
+                print("ERROR: unrecognized input --newdata=%s" % newdata_file)
+                sys.exit(1)
+        for ni, nfile in enumerate(nfiles): # iterate on files to predict
+            tdf = read_data(nfile, drops, scaler=scaler)
+            if  tcol in tdf.columns:
+                tdf = tdf.drop(tcol, axis=1)
+            if  verbose:
+                print("New data file", nfile)
+                print("Columns:", ','.join(tdf.columns))
+                print("test shapes:", tdf.shape)
+            datasets = [int(i) for i in list(tdf['dataset'])]
+            # 20160301 'dbs' header changed to 'dbsinst'
+            dfkeys = tdf.keys()
+            if  'dbsinst' in dfkeys:
+                dbs_h = 'dbsinst'
+            elif 'dbs' in dfkeys:
+                dbs_h = 'dbs'
+            else:
+                msg = "ERROR: file %s does not contain neither 'dbs' nor 'dbsinst' column"
+                print(msg % nfile)
+                sys.exit(1)
+            dbses = [int(i) for i in list(tdf[dbs_h])]
+            if  scaler:
+                tdf = getattr(preprocessing, scaler)().fit_transform(tdf)
+            if  verbose:
+                print(tdf)
+            time0 = time.time()
+            predictions = fit.predict(tdf) if not proba else np.asarray(fit.predict_proba(tdf))[:,list(fit.classes_).index(1)]
+            rtime += time.time()-time0
+            data = {'dataset':datasets, dbs_h: dbses, 'prediction':predictions}
+            out = pd.DataFrame(data=data)
```

Comment:
should be replaced with
out = pd.DataFrame({'dataset':datasets, dbs_h: dbses, 'prediction':predictions})

You created data variable and allocated memory for it, then just copy if to DataFrame. Instead, just create proper dataframe and don't waste program memory allocation for data. You don't even use data variable afterwards in this scope.